text/typeclass_control.md

```diff
@@ -0,0 +1,19 @@
+- Title: Typeclasses
+
+- Drivers: Cyril Cohen
+
+- Status: Draft
+
+----
+
+# Wishes
+
+- A cut tag between the (class) arguments of instances, with the same behavior as the prolog cut (as opposed to the Hint Cut). Syntax TBD
+E.g. (using "cut" as a placeholder for the future syntax)
+```coq
+Instance A `{C x} "cut" `{C' x y} : C (f x y) := ...
+```
+If a solution for (C x) is found, then no backtrack is allowed if a search for (C' x y) fails.
```

Comment:
If I understand correctly what you wish, it seems like this could also be expressed as an additional mode for `Hint Mode`, since it already specifies which arguments are input-only.

Reply:
As of today `Hint Mode` describes the status of the indexes of a **Class**.
"Cut" is a marker in a prolog clauses i.e. **Instances**, for which the number of arguments is independent with the number of indexes of a class.
So no, `Hint Mode` is not appropriate.

Reply:
It's really a prolog cut as in https://en.wikibooks.org/wiki/Prolog/Cuts_and_Negation

Reply:
FTR, you can encode a Prolog cut by directly writing a Hint Extern and using the `once` tactical over the set of typeclasses you want to cut on. It's very heavyweight though.

Reply:
I tried and failed, although I couldn't determine whether it was due to a mistake on my behalf or because the overhead is so heavy that it crashes... Debugging was so unreadable that I gave up.
Moreover, I am not sure what the interaction between unfolding and hints is.
If you have such an encoding, please share it with me so I can try yours :)

Reply:
@aspiwack Could we implement a cut on top of the engine actually, as a primitive ? Otherwise we need a kind of global setup for each instance, doing a kind of implicit "once" on the whole previous hypotheses

Reply:
However, my encoding had to be a bit more subtle than what you said because I do not want to cut arguments based on the typeclasses, but rather on the same typeclass but different positions of the arguments in an instance, so my encoding relies on an dummy typeclass with a Hint Extern which calls once typeclasses eauto...

Reply:
Here you go:
```coq
Class A := { a : nat }.
Class B := { b : nat }.

Arguments a _ : clear implicits.
Arguments b _ : clear implicits.

Class C := { ca : nat; cb : nat }.

Instance a₀ : A := { a := 0 }.
Instance a₁ : A := { a := 1 }.

Instance b₀ : B := { b := 0 }.
Instance b₁ : B := { b := 1 }.

Ltac instance :=
  (** Block instance resolution on C in the refine below *)
  change (id C);
  let x := fresh in
  unshelve refine (let x := _ in _); [exact A | | ];
  [once typeclasses eauto|];
  let av := eval unfold x in x in
  clear x;
  unshelve refine (let x := _ in _); [exact B | | ];
  [typeclasses eauto|];
  let bv := eval unfold x in x in
  clear x;
  refine (Build_C av.(@a) bv.(@b));
  change C
  .

Ltac list_instance A :=
  try (
    let x := fresh in
    unshelve refine (let x := _ in _); [exact A | | ];
    [typeclasses eauto|];
    let ans := eval unfold x in x in
    clear x;
    idtac ans;
    fail
  ).

Hint Extern 2 C => instance : typeclass_instances.

Goal True.
list_instance A.
list_instance B.
list_instance C.
```

Reply:
Another solution would be to have a modifier for using once on each argument but that's not as powerful as cut.

Reply:
Ok, I tried again and it almost work (without `once` though), and here I am:
```coq
(**************)
(* Prolog cut *) 
(**************)

Fact cut_key : unit. Proof tt.
Definition locked_with : unit -> forall {T}, T -> T := fun 'tt => @id.
Class cut (class : Type) := cut_field : locked_with cut_key class.

Section cut.
Context (class : Type).
Local Notation cut := (cut class).
Lemma cutE : cut = class.
Proof. now unfold cut; case cut_key. Qed.
Definition put_cut : cut -> class.
Proof. now rewrite cutE. Qed.
Definition get_cut : class -> cut.
Proof. now rewrite cutE. Qed.
End cut.

Notation "'cut c" := (@cut c) (at level 200).

Require Import Classes.Init.
Hint Extern 0 (@cut _) =>
  (apply (@get_cut _ _)) : typeclass_instances.

(* Testing prolog cut *)

Class C (n : nat) (m : nat).
Hint Mode C + - : typeclass_instances.
Instance a : C 0 1 | 1.
Instance b : C 0 0 | 2.
Instance d x y `{'cut C 0 x} `{C x y} : C 3 3.

Goal C 3 3.
Set Typeclasses Debug. 
Fail typeclasses eauto.
Abort.
```
The big problem here is that apply breaks the flow of the proof search and the traces start back to 1., which means that debugging becomes unreadable because of the interleaving between the numerous proof search that are printed in a random order. Moreover any options of typeclasses eauto are lost.

But if I replace `apply (@get_cut _ _)` by `once (class_apply get_cut)`, it does not work. The test which is supposed to fail (because the first solution that works is "a" and thus "b" should never be considered) now succeeds :(

Reply:
If I do this, it works now :
```coq
Hint Extern 0 (@cut _) =>
  (eapply @get_cut); once typeclasses eauto : typeclass_instances.
```

Reply:
@CohenCyril Yes, interleaved searches are a mess indeed, and a few times I've wondered how we could get more control in Hint Extern to control the backtracking of typeclasses. Maybe with Ltac2 we could get a handle on the continuations involved to express more elaborate control.